src/wallet/rpcwallet.cpp

```diff
@@ -1888,56 +1900,61 @@ UniValue listaccounts(const UniValue &params, bool fHelp)
             HelpExampleCli("listaccounts", "0") + "\nList account balances for 6 or more confirmations\n" +
             HelpExampleCli("listaccounts", "6") + "\nAs json rpc call\n" + HelpExampleRpc("listaccounts", "6"));
 
-    LOCK2(cs_main, pwalletMain->cs_wallet);
-
-    int nMinDepth = 1;
-    if (params.size() > 0)
-        nMinDepth = params[0].get_int();
-    isminefilter includeWatchonly = ISMINE_SPENDABLE;
-    if (params.size() > 1)
-        if (params[1].get_bool())
-            includeWatchonly = includeWatchonly | ISMINE_WATCH_ONLY;
-
     map<string, CAmount> mapAccountBalances;
-    for (const PAIRTYPE(CTxDestination, CAddressBookData) & entry : pwalletMain->mapAddressBook)
-    {
-        if (IsMine(*pwalletMain, entry.first, chainActive.Tip()) & includeWatchonly) // This address belongs to me
-            mapAccountBalances[entry.second.name] = 0;
-    }
 
-    for (map<uint256, CWalletTx>::iterator it = pwalletMain->mapWallet.begin(); it != pwalletMain->mapWallet.end();
-         ++it)
     {
-        const CWalletTx &wtx = (*it).second;
-        CAmount nFee;
-        string strSentAccount;
-        list<COutputEntry> listReceived;
-        list<COutputEntry> listSent;
-        int nDepth = wtx.GetDepthInMainChain();
-        if (wtx.GetBlocksToMaturity() > 0 || nDepth < 0)
-            continue;
-        wtx.GetAmounts(listReceived, listSent, nFee, strSentAccount, includeWatchonly);
-        mapAccountBalances[strSentAccount] -= nFee;
-        for (const COutputEntry &s : listSent)
+        // Locking cs_main ensures that the chain doesn't progress during our summation of balances.  This means that
+        // the balances will be consistent, although they may not point to the tip.  This API should report the tip
+        LOCK2(cs_main, pwalletMain->cs_wallet);
+
```

Comment:
why keep this lock on cs_main?

Reply:
I tried to explain in the comment.  If the lock is removed no crashes will happen but I believe that balances may be inconsistent if the blockchain is extended or rewound during the loop where the wallet evaluates its UTXOs.

For example, lets say we have 2 UTXOs 3 confirmations deep but the user is waiting for 4 confirmations.  The balance calculation loop skips the first one due to too few conf.  Next a block comes in.  The balance loop continues and then hits the 2nd one.  It sees that as having 4 confirms.  

So the reported balance is not correct for either blockchain state but reflects a state half-way between.  By taking cs_main, I freeze blockchain updates.  This means that the balance will be consistent with some blockchain height, albeit an outdated one.